plugins/linux-capture/xcomposite-helpers.c

```diff
@@ -0,0 +1,637 @@
+		XVisualInfo *visual = glXGetVisualFromFBConfig(disp, config);
+		if (!visual)
+			continue;
+
+		found = depth == visual->depth;
+		XFree(visual);
+		if (found)
+			break;
+	}
+	XFree(configs);
+	if (!found) {
+		blog(log_level, "no matching fb config found");
+		s->pixmap = 0;
+		return;
+	}
+
+	// Should be consistent format with config we are using. Since we searched on RGBA let's use RGBA here.
+	const int pixmap_attrs[] = {GLX_TEXTURE_TARGET_EXT, GLX_TEXTURE_2D_EXT,
+				    GLX_TEXTURE_FORMAT_EXT,
+				    GLX_TEXTURE_FORMAT_RGBA_EXT, None};
+
+	// Try very hard to capture errors in glXCreatePixmap for NVIDIA drivers
+	// where only one pixmap can be bound in GLX at a time.
+	pixmap_err = false;
+	XErrorHandler prev = XSetErrorHandler(catch_pixmap_errors);
+	s->glxpixmap = glXCreatePixmap(disp, config, s->pixmap, pixmap_attrs);
+	XSync(disp, false);
+
+	s->gltex = gs_texture_create(s->width, s->height, GS_RGBA_UNORM, 1, 0,
+				     GS_GL_DUMMYTEX);
+	GLuint gltex = *(GLuint *)gs_texture_get_obj(s->gltex);
+	glBindTexture(GL_TEXTURE_2D, gltex);
+	// Not respecting a captured glXCreatePixmap error will result in Xorg closing our connection.
+	if (!pixmap_err)
+		glXBindTexImageEXT(disp, s->glxpixmap, GLX_FRONT_EXT, NULL);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	// glxBindTexImageEXT might modify the textures format.
+	enum gs_color_format format = gs_format_from_tex();
+	// Check if texture is invalid... because X11 hates us.
+	int w;
+	int h;
+	glGetTexLevelParameteriv(GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &w);
+	glGetTexLevelParameteriv(GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &h);
+	glBindTexture(GL_TEXTURE_2D, 0);
+	// We must sync OBS texture format based on any glxBindTexImageEXT changes.
+	s->gltex->format = format;
+
+	XSync(disp, false);
+	if (pixmap_err || (uint32_t)w < s->width || (uint32_t)h < s->height) {
+		blog(log_level, "glXCreatePixmap failed: %s", pixmap_err_text);
+		glXDestroyPixmap(disp, s->glxpixmap);
+		XFreePixmap(disp, s->pixmap);
+		gs_texture_destroy(s->gltex);
+		s->pixmap = 0;
+		s->glxpixmap = 0;
+		s->gltex = 0;
+		XSetErrorHandler(prev);
+		return;
+	}
+	XSetErrorHandler(prev);
+
+	s->tex = gs_texture_create(
+		s->width - s->crop_left - s->crop_right - 2 * s->border,
+		s->height - s->crop_top - s->crop_bot - 2 * s->border, format,
+		1, 0, GS_GL_DUMMYTEX);
+	if (s->swapRedBlue) {
+		GLuint tex = *(GLuint *)gs_texture_get_obj(s->tex);
+		glBindTexture(GL_TEXTURE_2D, tex);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_B, GL_RED);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_R, GL_BLUE);
+		glBindTexture(GL_TEXTURE_2D, 0);
+	}
+}
+
+struct reg_item {
+	struct xcompcap *src;
+	xcb_window_t win;
+};
+static DARRAY(struct reg_item) watcher_registry = {0};
+static pthread_mutex_t watcher_lock = PTHREAD_MUTEX_INITIALIZER;
+
+void watcher_register(xcb_connection_t *conn, struct xcompcap *s)
```

Comment:
We depend on knowing the struct to signal the render thread via the `window_changed` attribute. This is why we use the struct pointer, so we have access to the shared state to signal.

We could layer up the abstraction and create a watcher_id+*bool and pass those around but it seemed excessive when a simple pointer would do.

Reply:
I'm trying my best not to suggest moving this back to xcomposite-input.c since you expressed the wish to keep it in a separate file, but it's getting increasingly hard for me to justify to myself it's worth keeping this split. It seems like the watcher routines are tightly locked to the xcompcap structure, and thus should be together. Apologies for the back and forth with the code :slightly_frowning_face: